usr.sbin/sysupgrade/sysupgrade.8
.\"	$OpenBSD: sysupgrade.8,v 1.22 2024/10/25 03:42:06 deraadt Exp $
.\"
.\" Copyright (c) 2019 Florian Obser <florian@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: October 25 2024 $
.Dt SYSUPGRADE 8
.Os
.Sh NAME
.Nm sysupgrade
.Nd upgrade system to the next release or a new snapshot
.Sh SYNOPSIS
.Nm
.Op Fl fkn
.Op Fl A | s
.Op Fl b Ar base-directory
.Op Fl R Ar version
.Op Ar installurl | path
.Sh DESCRIPTION
.Nm
is a utility to upgrade
.Ox
to a new release or snapshot if available.
.Pp
.Nm
downloads the necessary files to
.Pa /home/_sysupgrade ,
verifies them with
.Xr signify 1 ,
and copies bsd.rd to
.Pa /bsd.upgrade .
.Pp
.Nm
by default then reboots the system.
The bootloader will automatically choose
.Pa /bsd.upgrade ,
triggering a one-shot upgrade using the files in
.Pa /home/_sysupgrade .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A
Upgrade to the next Appli baseline.
.It Fl b Ar base-directory
Download files to
.Ar base-directory Ns / Ns Pa _sysupgrade
instead of
.Pa /home/_sysupgrade .
.It Fl f
For snapshots, force an already applied upgrade.
This option has no effect on releases.
.It Fl k
Keep the files in
.Pa /home/_sysupgrade .
By default they will be deleted after the upgrade.
.It Fl n
Fetch and verify the files and create
.Pa /bsd.upgrade
but do not reboot.
.It Fl R Ar version
Upgrade to a specific release version.
Only upgrades from one version to the next are tested.
Skipping versions may work.
Downgrading is unlikely to work.
.It Fl s
Upgrade to a snapshot.
The default is to upgrade to the next release.
.El
.Pp
When updating to a release or snapshot which lacks the required signify
keys in
.Pa /etc/signify ,
the missing keys will be downloaded in a secure way.
In the usual case, the keys will already be present because
.Ox
releases ship with the current key, the next key, and a collection of
older keys.
.Pp
See
.Xr upgrade.site 5
for how to customize the upgrade process.
.Sh FILES
.Bl -tag -width "/auto_upgrade.conf" -compact
.It Pa /auto_upgrade.conf
Response file for the ramdisk kernel.
.It Pa /bsd.upgrade
The ramdisk kernel to trigger an unattended upgrade.
.It Pa /etc/installurl
.Ox
mirror top-level URL for fetching an upgrade.
.It Pa /home/_sysupgrade
Directory the upgrade is downloaded to.
.El
.Sh SEE ALSO
.Xr signify 1 ,
.Xr installurl 5 ,
.Xr upgrade.site 5 ,
.Xr autoinstall 8 ,
.Xr release 8 ,
.Xr sysmerge 8
.Sh HISTORY
.Nm
first appeared in
.Ox 6.6 .


usr.sbin/sysupgrade/sysupgrade.sh
#!/bin/ksh
#
# $OpenBSD: sysupgrade.sh,v 1.63 2026/09/20 02:23:36 dgl Exp $
#
# Copyright (c) 1997-2015 Todd Miller, Theo de Raadt, Ken Westerback
# Copyright (c) 2015 Robert Peichaer <rpe@openbsd.org>
# Copyright (c) 2016, 2017 Antoine Jacoutot <ajacoutot@openbsd.org>
# Copyright (c) 2019 Christian Weisgerber <naddy@openbsd.org>
# Copyright (c) 2019 Florian Obser <florian@openbsd.org>
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

set -e
umask 0022
export PATH=/usr/bin:/bin:/usr/sbin:/sbin

ARCH=$(uname -m)
SETSDIR=/home/_sysupgrade

err()
{
	echo "${0##*/}: ${1}" 1>&2
	return ${2:-1}
}

usage()
{
	echo "usage: ${0##*/} [-fkn] [-A | -s] [-b base-directory] [-R version] [installurl | path]" 1>&2
	return 1
}

unpriv()
{
	local _file _rc=0 _user=_syspatch

	if [[ $1 == -f ]]; then
		_file=$2
		shift 2
	fi
 	if [[ -n ${_file} ]]; then
		>${_file}
		chown "${_user}" "${_file}"
	fi
	(($# >= 1))

	eval su -s /bin/sh ${_user} -c "'$@'" || _rc=$?

	[[ -n ${_file} ]] && chown root "${_file}"

	return ${_rc}
}

# Remove all occurrences of first argument from list formed by the remaining
# arguments.
rmel() {
	local _a=$1 _b _c

	shift
	for _b; do
		[[ $_a != "$_b" ]] && _c="${_c:+$_c }$_b"
	done
	echo -n "$_c"
}

SNAP=false
FILE=false
BASELINE=false
FORCE=false
FORCE_VERSION=false
KEEP=false
REBOOT=true
WHAT='release'

VERSION=$(uname -r)
NEXT_VERSION=$(echo ${VERSION} + 0.1 | bc)

while getopts Ab:fknrR:s arg; do
	case ${arg} in
	A)	BASELINE=true;;
	b)	SETSDIR=${OPTARG}/_sysupgrade;;
	f)	FORCE=true;;
	k)	KEEP=true;;
	n)	REBOOT=false;;
	r)	;;
	R)	FORCE_VERSION=true
		[[ ${OPTARG} == @([0-9]|[0-9][0-9]).[0-9] ]] ||
		    err "invalid version: ${OPTARG}"
		NEXT_VERSION=${OPTARG};;
	s)	SNAP=true;;
	*)	usage;;
	esac
done

(($(id -u) != 0)) && err "need root privileges"

if $SNAP && $BASELINE; then
	usage
fi

# XXX still needed?
set -A _KERNV -- $(sysctl -n kern.version |
	sed 's/^OpenBSD \([1-9][0-9]*\.[0-9]\)\([^ ]*\).*/\1 \2/;q')

BL=$(($(echo ${_KERNV[0]} | tr -d .) - 37))
NEXT_BL=$(($(echo $NEXT_VERSION | tr -d .) - 37))

shift $(( OPTIND -1 ))

case $# in
0)	MIRROR=$(while read _line; do _line=${_line%%#*}; [[ -n ${_line} ]] &&
	    print -r -- "${_line}" | grep -v '[[:cntrl:]"$&;<>\`|'\']
	    done </etc/installurl | tail -1) 2>/dev/null
	    [[ -z $MIRROR ]] && MIRROR=https://cdn.openbsd.org/pub/OpenBSD
	;;
1)	MIRROR=$1
	;;
*)	usage
esac
[[ $MIRROR == @(file|ftp|http|https)://* ]] ||
	FILE=true
$FORCE_VERSION && $SNAP &&
	err "incompatible options: -s -R $NEXT_VERSION"
$FORCE && ! $SNAP &&
	err "incompatible options: -f without -s"

USED_USR=$(df -P /usr | awk 'NR==2 {print substr($5, 1, length($5)-1)}')
if (( ${USED_USR} >= 90 )) && ! $FORCE; then
    err "/usr appears too small. See sysupgrade(8) manpage about PRUNING."
fi

if $SNAP; then
	WHAT='snapshot'
elif $BASELINE; then
	WHAT='baseline'
fi

if $SNAP || $BASELINE; then
	if $SNAP; then
		URL=${MIRROR}/snapshots/${ARCH}/
	elif $FORCE; then
		URL=${MIRROR}/bl$BL/${ARCH}/
	else
		URL=${MIRROR}/bl$NEXT_BL/${ARCH}/
	fi
else
	URL=${MIRROR}/${NEXT_VERSION}/${ARCH}/
	$FORCE_VERSION || ALT_URL=${MIRROR}/${VERSION}/${ARCH}/
fi

# Oh wait, this is a path install
if $FILE; then
	URL=file://$MIRROR/
	ALT_URL=
fi

install -d -o 0 -g 0 -m 0755 ${SETSDIR}
cd ${SETSDIR}

echo "Fetching from ${URL}"
if ! unpriv -f SHA256.sig ftp -N sysupgrade -Vmo SHA256.sig ${URL}SHA256.sig; then
	if [[ -n ${ALT_URL} ]]; then
		echo "Fetching from ${ALT_URL}"
		unpriv -f SHA256.sig ftp -N sysupgrade -Vmo SHA256.sig ${ALT_URL}SHA256.sig
		URL=${ALT_URL}
		NEXT_VERSION=${VERSION}
	else
		exit 1
	fi
fi

# The key extracted from SHA256.sig must precisely match a pattern
KEY=$(head -1 < SHA256.sig | cut -d' ' -f5 | \
	egrep '^openbsd-[[:digit:]]{2,3}-base.pub$' || true)
if [[ -z $KEY ]]; then
	echo "Invalid SHA256.sig file"
	exit 1
fi

if !$BASELINE; then
	# If required key is not in the system, get it from a signed bundle
	if ! [[ -r /etc/signify/$KEY ]]; then
		HAVEKEY=$(cd /etc/signify && ls -1 openbsd-*-base.pub | \
		    tail -2 | head -1 | cut -d- -f2)
		BUNDLE=sigbundle-${HAVEKEY}.tgz
		FWKEY=$(echo $KEY | sed -e 's/base/fw/')
		echo "Adding missing keys from bundle $BUNDLE"
		unpriv -f ${BUNDLE} ftp -N sysupgrade -Vmo $BUNDLE https://ftp.openbsd.org/pub/OpenBSD/signify/$BUNDLE
		signify -Vzq -m - -x $BUNDLE | (cd /etc/signify && tar xfz - $KEY $FWKEY)
		rm $BUNDLE
	fi

	unpriv -f SHA256 signify -Ve -x SHA256.sig -m SHA256
else
	_KEY=baseline-$BL-base.pub
	_NEXTKEY=baseline-$NEXT_BL-base.pub

	read _LINE <SHA256.sig
	case ${_LINE} in
	*\ ${_KEY})	SIGNIFY_KEY=/etc/signify/${_KEY} ;;
	*\ ${_NEXTKEY})	SIGNIFY_KEY=/etc/signify/${_NEXTKEY} ;;
	*)		err "invalid signing key" ;;
	esac

	[[ -f ${SIGNIFY_KEY} ]] || err "cannot find ${SIGNIFY_KEY}"

	unpriv -f SHA256 signify -Ve -p "${SIGNIFY_KEY}" -x SHA256.sig -m SHA256
fi
rm SHA256.sig

if cmp -s /var/db/installed.SHA256 SHA256 && ! $FORCE; then
	echo "Already on latest ${WHAT}."
	exit 0
fi

unpriv -f BUILDINFO ftp -N sysupgrade -Vmo BUILDINFO ${URL}BUILDINFO
unpriv cksum -qC SHA256 BUILDINFO

if [[ -e /var/db/installed.BUILDINFO ]]; then
	installed_build_ts=$(cut -f3 -d' ' /var/db/installed.BUILDINFO)
	build_ts=$(cut -f3 -d' ' BUILDINFO)
	if (( $build_ts <= $installed_build_ts )) && ! $FORCE; then
		echo "Downloaded ${WHAT} is older than installed system. Use -f to force downgrade."
		exit 1
	fi
fi

# INSTALL.*, bsd*, *.tgz
SETS=$(sed -n -e 's/^SHA256 (\(.*\)) .*/\1/' \
    -e '/^INSTALL\./p;/^bsd/p;/\.tgz$/p' SHA256)

OLD_FILES=$(ls)
OLD_FILES=$(rmel SHA256 $OLD_FILES)
DL=$SETS

[[ -n ${OLD_FILES} ]] && echo Verifying old sets.
for f in ${OLD_FILES}; do
	if cksum -C SHA256 $f >/dev/null 2>&1; then
		DL=$(rmel $f ${DL})
		OLD_FILES=$(rmel $f ${OLD_FILES})
	fi
done

[[ -n ${OLD_FILES} ]] && rm ${OLD_FILES}
for f in ${DL}; do
	unpriv -f $f ftp -N sysupgrade -Vmo ${f} ${URL}${f}
done

if [[ -n ${DL} ]]; then
	echo Verifying sets.
	unpriv cksum -qC SHA256 ${DL}
fi

cat <<__EOT >/auto_upgrade.conf
Location of sets = disk
Pathname to the sets = ${SETSDIR}/
Directory does not contain SHA256.sig. Continue without verification = yes
__EOT

if ! ${KEEP}; then
	CLEAN=$(echo BUILDINFO SHA256 ${SETS} | sed -e 's/ /,/g')
	cat <<__EOT > /etc/rc.firsttime
rm -f ${SETSDIR}/{${CLEAN}}
__EOT
fi

echo Fetching updated firmware.
set -A _NEXTKERNV -- $(what bsd |
	sed -n '2s/^[[:blank:]]OpenBSD \([1-9][0-9]*\.[0-9]\)\([^ ]*\).*/\1 \2/p')

if [[ ${_NEXTKERNV[1]} == '-current' ]]; then
	FW_URL=http://firmware.openbsd.org/firmware/snapshots/
else
	FW_URL=http://firmware.openbsd.org/firmware/${_NEXTKERNV[0]}/
fi
VNAME="${_NEXTKERNV[0]}" fw_update -p ${FW_URL} || true

install -F -m 700 bsd.rd /bsd.upgrade
logger -t sysupgrade -p kern.info "installed new /bsd.upgrade. Old kernel version: $(sysctl -n kern.version)"
sync

if ${REBOOT}; then
	echo Upgrading.
	exec reboot
else
	echo "Will upgrade on next reboot"
fi
